When new timeline is created, don't wait for compaction.

hlinnaka opened the pull request:
When a new root timeline is created, we want to flush all the data to disk before we return success to the caller. We were using checkpoint(CheckpointConfig::Forced) for that, but that also performs compaction. With the default settings, compaction will have no work after we have imported an empty database, as the image of that is too small to require compaction. However, with very small checkpoint_distance and compaction_target_size, compaction will run, and it can take a while.

PR #2785 adds new tests that use very small checkpoint_distance and compaction_target_size settings, and the test sometimes failed with "operation timed out" error in the client, when the create_timeline step took too long.

hlinnaka commented:
I'm not 100% sure if this fixes the "operation timed out" errors we've been seeing with PR #2785. But this seems like the right thing to do, even if there's more to those "operation timed out" errors than this.

problame reviewed (commented):
> PR https://github.com/neondatabase/neon/pull/2785 adds new tests that use very small checkpoint_distance and compaction_target_size settings, and the test sometimes failed with "operation timed out" error in the client, when the create_timeline step took too long.

Were we hitting the general test timeout there? Isn't that, like, 60 seconds or so?
If you're referring to the tests that I added to `test_remote_storage.py`: we're generating nowhere near enough data to keep compaction busy for that long.data to keep us busy for that long. Do we?!

hlinnaka commented:
> Were we hitting the general test timeout there? Isn't that, like, 60 seconds or so? If you're referring to the tests that I added to `test_remote_storage.py`: we're generating nowhere near enough data to keep compaction busy for that long.data to keep us busy for that long. Do we?!

No, it's a timeout from `reqwest`, the HTTP client. It's 30 s. If the client doesn't receive a response to the HTTP request in 30 s, you get the "operation timed out" error.  There is no error in the pageserver log. As far as the server is concerned, the operation succeded.

We could make the timeout longer, too, if we had to.

hlinnaka commented:
> Were we hitting the general test timeout there? Isn't that, like, 60 seconds or so?
> If you're referring to the tests that I added to `test_remote_storage.py`: we're generating nowhere near enough data to keep compaction busy for that long.data to keep us busy for that long.

It's surprisingly heavy. You set the target file size settings (checkpoint_distance and compaction_target_size) to 32 kB, so it splits all the data into 32 kB files. And there's at least 16 kB of overhead in each file too, for the header and the little index in the file. The test generated over 1600 files altogether.

I changed those settings to 64 kB in `test_timeline_deletion_with_files_stuck_in_upload_queue`; that's where I saw the error first. With that you get about 500 files, which is still a lot but manageable. I just noticed that there was another test, `test_remote_storage_upload_queue_retries`, which also uses 32 kB currently. Should probably bump that up to 64 kB too.

problame reviewed (approved):
Ok. Not sure whether we're just kicking the can down the road (*) but this change seems like the right thing to do, generally.

(*) = the place where we call the on-demand compaction HTTP handler
